fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from collections import OrderedDict
+import itertools
+from typing import TYPE_CHECKING, Callable, Dict, List, Tuple, Type, Union
+
+import torch
+from torch import Tensor
+import torch.distributed.rpc as rpc
+import torch.nn as nn
+
+Tensors = Tuple[Tensor, ...]
+TensorOrTensors = Union[Tensor, Tensors]
+
+LayerSpec = Tuple[str, Type[nn.Module], Tuple, Dict]
+
+if TYPE_CHECKING:
+    Module = nn.Module[TensorOrTensors]
+else:
+    Module = nn.Module
+
+BOUNCE_TENSORS = True
+
+
+def _verify_module(module: List[LayerSpec]) -> None:
+    if not isinstance(module, List):
+        raise TypeError("module must be a list")
+
+    for elem in module:
+        if not isinstance(elem, tuple):
+            raise TypeError("module must be a list of tuple")
+        if len(elem) != 4:
+            raise TypeError("each module tuple must contain (name, nn.module, args, kwargs)")
+        name, layer, args, kwargs = elem
+        if not (
+            isinstance(name, str)
+            and issubclass(layer, nn.Module)
+            and isinstance(args, tuple)
+            and isinstance(kwargs, dict)
+        ):
+            raise TypeError("each module tuple must contain (name, nn.module, args, kwargs)")
+
+
+class _ToHere(Module):
+    def __init__(self, device: str):
+        super().__init__()
+        self.device = device
+
+    def forward(self, x_rref: rpc.RRef) -> Tensor:  # type: ignore
+        if BOUNCE_TENSORS:
+            return x_rref.remote().cpu().to_here().to(self.device)
```

Comment:
I was hoping to understand this better, this moves the tensor to the cpu first?

Reply:
rpc (until v1.8.0) only supported cpu tensors so I had to bounce to cpu first. I recently learned of GPU support in v1.8.0 which I am now trying to incorporate.